Add drag-and-drop support.

danpark-salesforce opened the pull request:
This PR adds drag-and-drop support, leveraging new additions to the CDP Input domain (Input.setInterceptDrags, Input.dispatchDragEvent, and Input.dragIntercepted).                                            

Fixes #1376

danpark-salesforce committed:
Add drag-and-drop support.

This commit adds drag-and-drop support, leveraging new additions to the CDP
Input domain (Input.setInterceptDrags, Input.dispatchDragEvent, and
Input.dragIntercepted).

danpark-salesforce commented:
Hello! I've been trying to test drag-and-drop at work using a variety of E2E tools like Cypress and Webdriver. Unfortunately like everyone else I kept running into a frustrating issue where the thing getting dragged would just stick to my mouse cursor instead of getting dropped at a given location.

Luckily, it turns out that Chromium recently added some new hooks for dragging and I'm happy to report they work great! 

I know this PR is a bit light on the tests, but I wanted to make sure this was going in the right direction before I spent more time. Looking forward to your feedback, thanks!

mathiasbynens commented on test/drag-and-drop.spec.ts:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2018 Google Inc. All rights reserved.
```
```suggestion
 * Copyright 2021 Google Inc. All rights reserved.
```

mathiasbynens commented:
@jschfflr Could you PTAL (and consider the record/replay perspective)?

jschfflr reviewed (changes requested):
First of all, the general direction of this change looks good to me. Thanks for taking the time to create this patch!
I just have some small questions:
I'm not sure why you chose to create a new CDP Session for the drag events - could you elaborate on that or just use the already existing one?
Also, there is the optional `modifiers` flag - could you make sure that the current set of modifiers from the keyboard is respected?
For the return value of the drag: WDYT about returning the DragData directly (instead of the event)?

danpark-salesforce committed:
Address feedback.

* Update date in test
* Pass modifiers when dispatching drag events
* Return DragData directly instead of event

danpark-salesforce commented:
Thanks @jschfflr for taking a look! Updated the PR based on your feedback.

RE: `I'm not sure why you chose to create a new CDP Session for the drag events - could you elaborate on that..?`

This was for isolating the drag interception in case something else was also listening for drag events on the existing client. If this isn't a concern, I can use the existing client.

On a different note, the `drop()` method is currently dispatching three drag events in sequence (dragEnter, dragOver, and drop). I was thinking it might be better to move the dragEnter and dragOver into separate methods. What are your thoughts?

jschfflr commented:
Thanks for updating the patch!
I'd prefer to reuse the existing client but you have a good point about the isolation. 
WDYT about approaching this in a similar way as the network interception by exposing a `page.setDragInterception(true)`?

If you feel like the drop events would also be useful on their own, feel free to also expose them. 
But I'd prefer to have a single `drop` method because developers might not know about the specific event order for drag'n'drop events.
Also: WDYT about exposing a single `dragAndDrop` function that takes two selectors and drags one of them into the other?

danpark-salesforce committed:
Address feedback.

* Moved drag interception code to Page
* Added tests for drag interception enablement

danpark-salesforce commented:
Updated and moved the drag interception to the Page class; thanks for pointing me to that.

>developers might not know about the specific event order for drag'n'drop events.

Makes total sense, will keep the drop method as is. ~I'll also add methods for the other drag events in case things like dragover effects need to be tested.~ (edit: I think adding a delay parameter to the `drop()` would be simpler).

>Also: WDYT about exposing a single dragAndDrop function that takes two selectors and drags one of them into the other?

I think a `dragAndDrop` function is a good idea. One issue I do see with using selectors is it won't work with web components / shadow DOM. What are your thoughts on passing either selectors or element handles? Also, if we go down this route, we could add the `dragAndDrop` method directly to the Input as well.

jschfflr commented:
Awesome, thanks! 
I think, having individual methods for each event could still be valuable if a user only wants to test one of them.
But the delay parameter sounds super helpful too. Could you add both?

We do have the `pierce/` query handler to deal with selectors that use the shadow dom. But I agree with you that using element handles is a better solution. Let's use them instead of the selectors.

jschfflr commented:
@danparksf Hi, did you already have time to look into the changes we discussed? 
I'd love to get this landed soon :)

danpark-salesforce commented:
Hi @jschfflr, thanks for the feedback! Been meaning to look into updating this PR and finally have some time to work on them. Should have an update later today, thanks!

danpark-salesforce committed:
Address feedback.

* Add separate methods for dragEnter and dragOver
* Fix lint errors
* Add some docs for ElementHandle.drag() and Mouse.drag()

danpark-salesforce commented:
@jschfflr haven't added the `dragAndDrop()` yet since I wanted to check something with you. In order to drag to a target element, we need to get the x, y coordinates of the target element. At the moment there is a method `_clickablePoint()` that can do that for us. Are you fine with having a public method to expose it?

jschfflr commented:
@danparksf Yes, feel free to mark this function as public!

jschfflr commented:
Maybe you can also change the return type and expose a public `Point` type similar to `BoundingBox`

danpark-salesforce committed:
Add dragAndDrop method.

Also update API docs.

danpark-salesforce commented:
Thanks for the tips, PR updated!

danpark-salesforce committed:
Fix lint errors.

jschfflr commented:
@danparksf Not sure if you are already aware of this, but you can use `npm run lint` locally to find all the lint errors that will also occur on the bots later.

danpark-salesforce committed:
fix: lint error

danpark-salesforce commented:
Thanks, should be passing the lint check now.

jschfflr commented:
@danparksf Thanks! The linter is happy now. Could you also look into the failing unit tests? You can run them locally with `npm run unit`

danpark-salesforce commented:
@jschfflr looks like you fixed the tests.. thanks for doing that!

jschfflr commented on docs/api.md:
```diff
@@ -2960,6 +2986,62 @@ Shortcut for [`mouse.move`](#mousemovex-y-options), [`mouse.down`](#mousedownopt
+  - `y` <[number]> y coordinate
+- returns: <[Promise<[DragData]>]>
+
+This method creates and captures a dragevent from a given point.
```
Can you explain what target is used for in this context?

jschfflr commented on docs/api.md:
```diff
@@ -4045,6 +4131,45 @@ If the element is detached from DOM, the method throws an error.
+
+#### elementHandle.dragAndDrop(target[, options])
+
+- `target` <[Object]>
```
Can you change this to `ElementHandle`?

jschfflr reviewed (commented):
Looks great, thanks!

danpark-salesforce committed:
updated docs

danpark-salesforce committed:
run drag-and-drop tests for chrome only

jschfflr reviewed (approved):
LGTM, thanks! Let's get this landed! 🎉

jschfflr committed:
Merge branch 'main' into drag-and-drop

jschfflr committed:
chore: run eslint

jschfflr commented:
You can fix the lint errors by running `npm run eslint-fix`.
And maybe you can change the tests back to `it` and use `describeChromeOnly` for the whole block instead.

jschfflr commented:
@danparksf Sorry for the confusion - github didn't show me the commits I pushed and therefore I thought it didn't work. 
I hope it's ok for you that I fixed the eslint errors. If yes, I'm happy to land the change!

jschfflr committed:
Merge branch 'main' into drag-and-drop

danpark-salesforce commented:
@jschfflr no problem, thanks for the fix!

danpark-salesforce commented:
@googlebot I consent.

jschfflr committed:
Merge branch 'main' into drag-and-drop

jschfflr commented:
@googlebot I consent.

jschfflr commented:
@danparksf Awesome, let's get this landed then :)

jschfflr commented:
@danparksf Congrats on getting this landed!

danpark-salesforce commented:
Thanks @jschfflr for your help!

mathiasbynens commented on src/common/Page.ts:
```diff
@@ -748,6 +749,10 @@ export class Page extends EventEmitter {
     return this._accessibility;
   }
 
+  get isDragInterceptionEnabled(): boolean {
```
It's surprising/unexpected that this is a getter. Other `page.is*` APIs are methods. Patch: https://github.com/puppeteer/puppeteer/pull/7419

mathiasbynens commented on test/drag-and-drop.spec.ts:
```diff
@@ -0,0 +1,125 @@
+  describeChromeOnly,
+} from './mocha-utils'; // eslint-disable-line import/extensions
+
+describeChromeOnly('Input.drag', function () {
```
`page.isDragInterceptionEnabled` isn’t tested.

jschfflr replied:
Good catch, I'll add a test for it. Don't we have presubmit checks in place for this?

jschfflr replied:
I looks like this has already been addressed in https://github.com/puppeteer/puppeteer/pull/7419
